_posts/2018-10-22-story-about-one-small-library.md

```diff
@@ -0,0 +1,71 @@
+---
+layout:     post
+title:      Story about one small library
+date:       2018-10-15
+summary:    Ethereumex evolution
+categories: ethereum
+---
+
+![babel](https://i.imgur.com/0lgyBVW.jpg)
+
+### Introduction
+
+A couple of years ago I got interested in Ethereum. My favourite programming language at the time was Elixir (by the way, Elixir is still my favourite language). My first contribution to the Ethereum community was [Ethereumex](https://github.com/exthereum/ethereumex) - Elixir JSON-RPC client for the Ethereum.
+
+In this post, I'll describe its evolution. I have got a couple of other open source projects that I wrote or was a part of the team. But I want to describe the evolution of Ethereumex because this small library is a good example of a hobby project that was written on the weekend and then got relatively popular in the community. I won't go into too much detail about the technical details of Ethereumex as I want this post to be about the open source, not about technology that it implements.
+
+### History
+
+From the first days of the project, I tried to follow community guides of good open source projects. One of the things I did for this is I added `CHANGELOG.md` file to easier track changes between versions. I'll use this file to describe the development of Ethereumex.
+
+#### 0.1.x
+
+As described in the introduction, the first version (`0.1.0`) was written literally over the weekend at the beginning of the summer of 2017. Most of my open source work is done in my free time. And I'm sure I'm not alone at this.
+
+This version was pretty simple: JSON-RPC methods were generated by a macro. Unfortunately, automatic code generation has its downsides. Macros certainly shouldn't be used for defining public methods that a user will interact with. The only difference between these methods was their names. They all accepted a list of params.
+
+Surprisingly, some people used this version, I even received [one pull request](https://github.com/exthereum/ethereumex/pull/2) that was published in `0.1.1` version.
+
+#### 0.2.0 - 0.3.x
+
+In October of 2017, I decided to take a week-long holiday. My full-time job at the time was pretty exhausting: I was writing digital book/comic-book processing services in Java and Kotlin and physically didn't have time for anything else.
+
+On this holiday I rewrote Ethereumex from scratch. I got rid of automatic code generation, public API methods were defined using Elixir's behaviours. There is a [detailed post called "Elixir macros are evil (sometimes)"](https://www.badykov.com/elixir/2017/10/18/macros-are-evil/) about this rewrite on my blog. I was satisfied with what I ended up with. I even got a comment from Jose Valim (creator of Elixir). Read my post about rewrite if you're interested.
```

Comment:
Did you make the public API with Elixir's behaviours instead of using macros? If so, maybe say something like, "generation, using Elixir's behaviours instead" or something like that?